### PR TITLE
Material design color picker with Windows implementation

### PR DESCRIPTION
**Pull request by bleroy**
materialdesigncolorpicker

Known issues:

* [ ] Styling is not 100% the planned design from #136 (width of the popup doesn't grow when material design tab is selected, for instance)
* [ ] Tab icons are not final
* [x] There are binding errors popping up in the output window when running the standalone app
* [x] It takes two clicks for changes to fully register (first time, selected style may not apply, or selection may not be acknowledged; I believe the problem is in `ChoiceControl` when `SelectedItem` is bound rather than `SelectedValue`)
* [x] Black and White colors are not in the design. They should be, will follow-up on what the design should be for those.

### CODE REVIEW

File: Xamarin.PropertyEditing.Windows/Themes/Resources.xaml [hunk 1]
```diff
@@ -1113,6 +1190,150 @@
 		</Setter>
 	</Style>
 
+	<Style x:Key="BrushChoiceControlItem" BasedOn="{StaticResource {x:Type RadioButton}}" TargetType="RadioButton">
```
Comment (ermau) [inline Xamarin.PropertyEditing.Windows/Themes/Resources.xaml]: Can we name this more specifically? It's meant for the material picker right? It sounds like its for selecting the brush type.

**Comment by ermau** [comment]
>  It takes two clicks for changes to fully register (first time, selected style may not apply, or selection may not be acknowledged

This must be fixed.

File: Xamarin.PropertyEditing.Windows/BrushTabbedEditorControl.cs [hunk 2]
```diff
@@ -83,14 +115,43 @@ private void StorePreviousBrush ()
...
+				ShowSelectedTab ();
 				break;
 			case CommonSolidBrush _:
 				this.brushChoice.SelectedValue = solid;
+				ShowSelectedTab ();
+				break;
+			}
+		}
+
+		private void ShowSelectedTab()
```
[the remark this answers is not in the document]
Comment (bleroy) [reply Xamarin.PropertyEditing.Windows/BrushTabbedEditorControl.cs]: That will be part of the changes we talked about on the view model, I didn't attempt to fix that in this PR.

File: Xamarin.PropertyEditing/Drawing/CommonColor.cs [hunk 3]
```diff
@@ -43,6 +44,10 @@ public CommonColor (byte r, byte g, byte b, byte a = 255)
 		/// Blue component
 		/// </summary>
 		public byte B { get; }
+		/// <summary>
+		/// An optional label for the color, that does not affect equality or anything else.
```
[the remark this answers is not in the document]
Comment (bleroy) [reply Xamarin.PropertyEditing/Drawing/CommonColor.cs]: Good point.

File: Xamarin.PropertyEditing.Windows/Themes/Resources.xaml (hunk 1 above)
Comment (bleroy) [reply Xamarin.PropertyEditing.Windows/Themes/Resources.xaml]: Sure.

**Comment by bleroy** [comment]
Definitely. The rest as well.

File: Xamarin.PropertyEditing/ViewModels/MaterialDesignColorViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+				.OrderBy (palette => palette.Colors.Min (
+					paletteColor => CommonColor.SquaredDistance (paletteColor, Color)))
+				.First ();
+
+		public CommonColor? NormalColor
+		{
+			get {
+				if (this.normalColor.HasValue) return this.normalColor.Value;
+				MaterialColorScale scale = FindPalette (ColorName, false);
+				foreach (CommonColor normalColor in scale.Colors) {
```
Comment (lewing) [inline Xamarin.PropertyEditing/ViewModels/MaterialDesignColorViewModel.cs]: It looks like FindPalette can return null so this needs to handle that. Actually there are a lot of places where FindPalette results are accessed without a check?

Comment (bleroy) [reply Xamarin.PropertyEditing/ViewModels/MaterialDesignColorViewModel.cs]: Yes, you're right.

File: Xamarin.PropertyEditing/Drawing/CommonColor.cs [hunk 5]
```diff
@@ -8,12 +8,13 @@ namespace Xamarin.PropertyEditing.Drawing
 	[Serializable]
 	public struct CommonColor : IEquatable<CommonColor>
 	{
-		public CommonColor (byte r, byte g, byte b, byte a = 255)
+		public CommonColor (byte r, byte g, byte b, byte a = 255, string name = "")
```
[the remark this answers is not in the document]
Comment (bleroy) [reply Xamarin.PropertyEditing/Drawing/CommonColor.cs]: No particular reason, null works just as well. Changing it.

File: Xamarin.PropertyEditing/ViewModels/MaterialDesignColorViewModel.cs (hunk 4 above)
Comment (lewing) [reply Xamarin.PropertyEditing/ViewModels/MaterialDesignColorViewModel.cs]: 👍

**Pull request merged**